USER_MANUAL.md

```diff
@@ -1,12 +1,355 @@
+three kinds of declarations shown above.
+
+1.  In the first stage, system primitive declarations are gathered and a parser
+    is constructed.
+2.  In the second stage, the system primitive parser is used to parse
+    configuration declarations; in our BNF grammar below, we use `{TermBubble}`
+    to describe things which can be parsed using the primitive syntax.
+3.  Finally, a third stage parser is constructed containing both primitive and
+    configuration syntax is constructed which is used to parse contexts,
+    context aliases, and rules. In our BNF grammar below, we use the words
+    `{Bubble}` and `{BubbleWithRewrites}` to describe parsers which contain the
+    combined primitive and configuration syntax. The `{BubbleWithRewrites}`
+    category additional adds the rewrite arrow syntax (`=>`) needed for parsing
+    `rule`s and `context`s.
+
+**Note:** _the parsers listed above are all used internally by the K compiler
+`kompile`. As part of compilation, `kompile` will generate a separate
+language-specific parser that can only parse programs in the specified language
+syntax. The language-specific parser is the one invoked when using `krun`._
+
+We now present our high-level BNF syntax:
+
+```
+  File         ::= RequiresList ModuleList
+
+  RequiresList ::= "" | Requires RequiresList
+  Requires     ::= RequireCmd {String}
+  RequireCmd   ::= "requires" | "require"
+
+  Module       ::= "module" {ModuleId} AttributeSet ImportList SentenceList "endmodule"
+
+  AttributeSet ::= "" | "[" Attributes "]"
+  Attributes   ::= Attributes "," {Attribute}
+
+  ImportList   ::= "" | Import ImportList
+  Import       ::= ImportCmd {ModuleId}
+  ImportCmd    ::= "imports" | "import"
+
+  SentenceList ::= "" | Sentence SentenceList
+  Sentence     ::= SortDeclaration
+                 | GrammarDeclaration
+                 | LexicalDeclaration
+                 | AssociativityDeclaration
+                 | PriorityDeclaration
+                 | ConfigDeclaration
+                 | RuleDeclaration
+                 | ContextDeclaration
+                 | ContextAliasDeclaration
+```
+
+We can further drill down by sentence type.
+
+```
+  SortDeclaration    ::= "syntax" {SortId} AttributeSet
+
+  GrammarDeclaration ::= "syntax" {SortId} "::=" AlternativeList
+  AlternativeList    ::= Production
+                       | Production "|" AlternativeList
+
+  Production         ::= ProductionSyntax AttributeSet
+  ProductionSyntax   ::= FunctionSyntax
+                       | GenericSyntax
+
+  FunctionNotation   ::= {Id} "(" ParameterList ")"
+  ParameterList      ::= "" | NeParameterList
+  NeParameterList    ::= Parameter | Parameter "," NeParameterList
+  Parameter          ::= OptParamName {SortId}
+  OptParamName       ::= "" | {Id} ":"
+
+  GenericNotation    ::= StringOrSortIdList
+  StringOrSortIdList ::= StringOrSortId | StringOrSortId StringOrSortIdList
+  StringOrSortIdList ::= {String} | {SortId}
```

Comment:
this is also not correct as you are not allowing for regular expression terminals

Reply:
I removed this section from this PR because it needs more polish.